code/_onclick/click.dm

```diff
@@ -200,42 +200,79 @@
 	return FALSE
 
 /**
- * A backwards depth-limited breadth-first-search to see if the target is
- * logically "in" anything adjacent to us.
+ * Returns TRUE if a movable can "Reach" this atom. This is defined as adjacency
+ *
+ * Args:
+ * * user: The movable trying to reach us.
+ * * reacher_range: How far the reacher can reach.
+ * * depth: How deep nested inside of an atom contents stack an object can be.
+ * * direct_access: Do not override. Used for recursion.
  */
-/atom/proc/CanReach(atom/ultimate_target, obj/item/tool, view_only = FALSE)
-	var/list/direct_access = DirectAccess()
-	var/depth = 1 + (view_only ? STORAGE_VIEW_DEPTH : INVENTORY_DEPTH)
-
-	var/list/closed = list()
-	var/list/checking = list(ultimate_target)
+/atom/proc/IsReachableBy(atom/movable/user, reacher_range = 1, depth = INFINITY, direct_access = user.DirectAccess())
+	SHOULD_NOT_OVERRIDE(TRUE)
 
-	while (checking.len && depth > 0)
-		var/list/next = list()
-		--depth
+	if(isnull(user))
+		return FALSE
 
-		for(var/atom/target in checking)  // will filter out nulls
-			if(closed[target] || isarea(target))  // avoid infinity situations
-				continue
+	if(src in direct_access)
+		return TRUE
 
-			if(isturf(target) || isturf(target.loc) || (target in direct_access) || (ismovable(target) && target.flags_1 & IS_ONTOP_1) || target.loc?.atom_storage) //Directly accessible atoms
-				if(Adjacent(target) || (tool && CheckToolReach(src, target, tool.reach))) //Adjacent or reaching attacks
-					return TRUE
+	// This is a micro-opt, if any turf ever returns false from IsContainedAtomAccessible, change this.
+	if(isturf(loc) || isturf(src))
+		if(CheckReachableAdjacency(user, reacher_range))
+			return TRUE
 
-			closed[target] = TRUE
+	depth--
 
-			if (!target.loc)
-				continue
+	if(depth <= 0)
+		return FALSE
 
-			if(target.loc.atom_storage)
-				next += target.loc
+	if(isnull(loc) || isarea(loc) || !loc.IsContainedAtomAccessible(src, user))
+		return FALSE
 
-		checking = next
+	return loc.IsReachableBy(user, reacher_range, depth, direct_access)
 
-	if(SEND_SIGNAL(src, COMSIG_ATOM_CANREACH, ultimate_target) & COMPONENT_ALLOW_REACH)
+/atom/proc/CheckReachableAdjacency(atom/movable/reacher, reacher_range)
+	if(reacher.Adjacent(src))
 		return TRUE
 
-	return FALSE
+	if(isliving(reacher))
+		var/mob/living/living_reacher = reacher
+		if(living_reacher.reach_length > reacher_range)
+			reacher_range = living_reacher.reach_length
+
+	return (reacher_range > 1) && RangedReachCheck(reacher, src, reacher_range)
+
+/// Called by IsReachableBy() to check for ranged reaches.
+/proc/RangedReachCheck(atom/movable/here, atom/movable/there, reach)
+	if(!here || !there)
+		return FALSE
+
+	if(reach <= 1)
+		return FALSE
+
+	// Prevent infinite loop.
+	if(istype(here, /obj/effect/abstract/reach_checker))
+		return FALSE
+
+	var/obj/effect/abstract/reach_checker/dummy = new(get_turf(here))
```

Comment:
This is awful. Considering how frequently this proc gets called create and deleting temporary atoms may strain the gc and is a bad practice in general

Have you tried using the proc AbstractMove()? Maybe that's what you're looking for

Reply:
this proc is almost never called. this is only if you have a reach greater than 1 which only a single item in the game grants.

Reply:
It's good to know that the proc isn't hot but if there exists a better way to do it then we should try that